File: src/eYcel/transformations.py
"""
Transformation functions for encrypting and decrypting Excel data.

Each forward transform has a matching reverse so that encrypt → decrypt
produces the original value.  The only *irreversible* transform is `hash`
(one-way by design) and `anonymize` (random replacement).
"""
import hashlib
import random
import string
from typing import Any, Dict, Optional, Union
from datetime import datetime, date, timedelta


# ---------------------------------------------------------------------------
# Forward transforms (encrypt direction)
# ---------------------------------------------------------------------------

def transform_hash(value: Any, salt: str) -> str:
    """
    One-way SHA-256 hash of a value with a salt prefix.

    Args:
        value: Value to hash (converted to string).
        salt:  Per-column random salt string.

    Returns:
        12-character hex digest prefix (safe, short, readable).
    """
    raw = f"{salt}:{value}"
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()[:12]


def transform_offset_date(
    date_val: Union[datetime, date], offset_days: int
) -> Union[datetime, date]:
    """
    Shift a date/datetime by a fixed number of days.

    Args:
        date_val:    Original date or datetime.
        offset_days: Days to add (negative = subtract).

    Returns:
        Shifted date of the same type.
    """
    delta = timedelta(days=offset_days)
    return date_val + delta


def transform_offset_number(num_val: float, offset: float) -> float:
    """
    Add a fixed offset to a numeric value.

    Args:
        num_val: Original number.
        offset:  Value to add (can be negative).

    Returns:
        Shifted number (float).
    """
    return float(num_val) + float(offset)


def transform_scale(value: float, factor: float) -> float:
    """
    Multiply a numeric value by a secret factor.

    Args:
        value:  Original number.
        factor: Multiplication factor (must not be 0).

    Returns:
        Scaled number (float).

    Raises:
        ValueError: If factor is zero.
    """
    if factor == 0:
        raise ValueError("Scale factor must not be zero.")
    return float(value) * float(factor)


def transform_shuffle(value: str, mapping: Dict[str, str]) -> str:
    """
    Rename a category string using a pre-built mapping dictionary.

    Args:
        value:   Original category string.
        mapping: Dict mapping original → anonymised label.

    Returns:
        Anonymised label, or the original value if not found in mapping.
    """
    return mapping.get(str(value), str(value))


def transform_keep(value: Any) -> Any:
    """
    Pass-through — return value unchanged.

    Args:
        value: Any cell value.

    Returns:
        Identical value.
    """
    return value


def transform_anonymize(value: Any, col_type: str, _seed: Optional[int] = None) -> Any:
    """
    Replace a value with a realistic-looking fake of the same type.

    Args:
        value:    Original value (used only to determine length/shape).
        col_type: One of 'string', 'int', 'float', 'date', 'categorical'.
        _seed:    Optional RNG seed for deterministic tests.

    Returns:
        A plausible fake value of the same Python type.
    """
    rng = random.Random(_seed)
    if col_type == "int":
        return rng.randint(1, 99999)
    if col_type == "float":
        return round(rng.uniform(0.01, 99999.99), 2)
    if col_type == "date":
        base = date(2000, 1, 1)
        return base + timedelta(days=rng.randint(0, 9000))
    if col_type in ("string", "categorical"):
        length = max(4, len(str(value)))
        return "".join(rng.choices(string.ascii_uppercase, k=length))
    # fallback
    return value


# ---------------------------------------------------------------------------
# Reverse transforms (decrypt direction)
# ---------------------------------------------------------------------------

# ---------------------------------------------------------------------------
# Consistent text substitution
# ---------------------------------------------------------------------------

SYLLABLES = [
    "ba", "be", "bi", "bo", "bu", "da", "de", "di", "do", "du",
    "fa", "fe", "fi", "fo", "fu", "ga", "ge", "gi", "go", "gu",
    "ka", "ke", "ki", "ko", "ku", "la", "le", "li", "lo", "lu",
    "ma", "me", "mi", "mo", "mu", "na", "ne", "ni", "no", "nu",
    "pa", "pe", "pi", "po", "pu", "ra", "re", "ri", "ro", "ru",
    "sa", "se", "si", "so", "su", "ta", "te", "ti", "to", "tu",
    "va", "ve", "vi", "vo", "vu", "za", "ze", "zi", "zo", "zu",
]


def fake_word_from_index(index: int) -> str:
    """Generate a readable fake word from an index."""
    base = len(SYLLABLES)
    s1 = index % base
    s2 = (index // base) % base
    word = SYLLABLES[s1].capitalize() + SYLLABLES[s2]
    tier = index // (base * base)
    if tier > 0:
        word += str(tier)
    return word


def build_global_text_map(unique_texts: list) -> Dict[str, str]:
    """Build a global text substitution map. Same text → same fake word."""
    sorted_texts = sorted(set(str(t) for t in unique_texts if t))
    return {orig: fake_word_from_index(i) for i, orig in enumerate(sorted_texts)}


def substitute_text_in_formula(formula: str, text_map: Dict[str, str]) -> str:
    """Replace text literals inside a formula using the text map."""
    result = []
    i = 0
    chars = list(formula)
    while i < len(chars):
        if chars[i] == '"':
            start = i + 1
            i += 1
            while i < len(chars) and chars[i] != '"':
                i += 1
            literal = "".join(chars[start:i])
            replacement = text_map.get(literal, literal)
            result.append('"')
            result.append(replacement)
            result.append('"')
            if i < len(chars):
                i += 1
        else:
            result.append(chars[i])
            i += 1
    return "".join(result)


def reverse_text_in_formula(formula: str, inverse_map: Dict[str, str]) -> str:
    """Reverse text substitutions inside a formula."""
    return substitute_text_in_formula(formula, inverse_map)


# ---------------------------------------------------------------------------
# Reverse transforms (decrypt direction)
# ---------------------------------------------------------------------------

def reverse_offset_date(
    encrypted_date: Union[datetime, date], offset_days: int
) -> Union[datetime, date]:
    """
    Reverse a date offset by subtracting the original shift.

    Args:
        encrypted_date: The shifted date.
        offset_days:    The same offset_days used during encryption.

    Returns:
        Original date.
    """
    return encrypted_date - timedelta(days=offset_days)


def reverse_offset_number(encrypted_val: float, offset: float) -> float:
    """
    Reverse a number offset by subtracting the original shift.

    Args:
        encrypted_val: The shifted number.
        offset:        The same offset used during encryption.

    Returns:
        Original number (float).
    """
    return float(encrypted_val) - float(offset)


def reverse_scale(encrypted_val: float, factor: float) -> float:
    """
    Reverse scaling by dividing.

    Args:
        encrypted_val: The scaled number.
        factor:        The same factor used during encryption.

    Returns:
        Original number (float).

    Raises:
        ValueError: If factor is zero.
    """
    if factor == 0:
        raise ValueError("Scale factor must not be zero.")
    return float(encrypted_val) / float(factor)


def reverse_shuffle(encrypted_val: str, mapping: Dict[str, str]) -> str:
    """
    Reverse a shuffle by inverting the mapping.

    Args:
        encrypted_val: The anonymised category label.
        mapping:       The original original → anonymised mapping.

    Returns:
        Original category string, or the encrypted value if not found.
    """
    inverse = {v: k for k, v in mapping.items()}
    return inverse.get(str(encrypted_val), str(encrypted_val))

File: eYcel-wasm/src/transformations.rs
use chrono::{NaiveDate, NaiveDateTime, TimeDelta};
use rand::Rng;
use sha2::{Digest, Sha256};
use std::collections::HashMap;

use crate::cell_value::CellValue;

// ---------------------------------------------------------------------------
// Forward transforms (encrypt direction)
// ---------------------------------------------------------------------------

/// One-way SHA-256 hash with salt prefix. Returns 12-char hex digest.
pub fn transform_hash(value: &str, salt: &str) -> String {
    let raw = format!("{}:{}", salt, value);
    let hash = Sha256::digest(raw.as_bytes());
    format!("{:x}", hash)[..12].to_string()
}

/// Shift a NaiveDate by a fixed number of days.
pub fn transform_offset_date(date_val: NaiveDate, offset_days: i32) -> NaiveDate {
    date_val + TimeDelta::days(offset_days as i64)
}

/// Shift a NaiveDateTime by a fixed number of days.
pub fn transform_offset_datetime(dt: NaiveDateTime, offset_days: i32) -> NaiveDateTime {
    dt + TimeDelta::days(offset_days as i64)
}

/// Add a fixed offset to a numeric value.
pub fn transform_offset_number(num_val: f64, offset: f64) -> f64 {
    num_val + offset
}

/// Multiply a numeric value by a factor.
pub fn transform_scale(value: f64, factor: f64) -> Result<f64, &'static str> {
    if factor == 0.0 {
        return Err("Scale factor must not be zero.");
    }
    Ok(value * factor)
}

/// Rename a category using a mapping dict. Returns original if not found.
pub fn transform_shuffle(value: &str, mapping: &HashMap<String, String>) -> String {
    mapping.get(value).cloned().unwrap_or_else(|| value.to_string())
}

/// Pass-through — return value unchanged.
pub fn transform_keep(value: CellValue) -> CellValue {
    value
}

/// Replace value with a realistic fake of the same type. Irreversible.
pub fn transform_anonymize(value: &CellValue, col_type: &str, rng: &mut impl Rng) -> CellValue {
    match col_type {
        "int" => CellValue::Int(rng.gen_range(1..=99999)),
        "float" => {
            let v: f64 = rng.gen_range(0.01..99999.99);
            CellValue::Float((v * 100.0).round() / 100.0)
        }
        "date" => {
            let base = NaiveDate::from_ymd_opt(2000, 1, 1).unwrap();
            let days = rng.gen_range(0..9000);
            CellValue::Date(base + TimeDelta::days(days))
        }
        "string" | "categorical" => {
            let length = match value {
                CellValue::String(s) => s.len().max(4),
                _ => 4,
            };
            let s: String = (0..length)
                .map(|_| (b'A' + rng.gen_range(0..26)) as char)
                .collect();
            CellValue::String(s)
        }
        _ => value.clone(),
    }
}

// ---------------------------------------------------------------------------
// Reverse transforms (decrypt direction)
// ---------------------------------------------------------------------------

/// Reverse a date offset by subtracting the original shift.
pub fn reverse_offset_date(encrypted_date: NaiveDate, offset_days: i32) -> NaiveDate {
    encrypted_date - TimeDelta::days(offset_days as i64)
}

/// Reverse a datetime offset.
pub fn reverse_offset_datetime(dt: NaiveDateTime, offset_days: i32) -> NaiveDateTime {
    dt - TimeDelta::days(offset_days as i64)
}

/// Reverse a number offset by subtracting.
pub fn reverse_offset_number(encrypted_val: f64, offset: f64) -> f64 {
    encrypted_val - offset
}

/// Reverse scaling by dividing.
pub fn reverse_scale(encrypted_val: f64, factor: f64) -> Result<f64, &'static str> {
    if factor == 0.0 {
        return Err("Scale factor must not be zero.");
    }
    Ok(encrypted_val / factor)
}

/// Reverse a shuffle by inverting the mapping.
pub fn reverse_shuffle(encrypted_val: &str, mapping: &HashMap<String, String>) -> String {
    let inverse: HashMap<&str, &str> = mapping.iter().map(|(k, v)| (v.as_str(), k.as_str())).collect();
    inverse
        .get(encrypted_val)
        .map(|s| s.to_string())
        .unwrap_or_else(|| encrypted_val.to_string())
}

// ---------------------------------------------------------------------------
// Global text substitution (consistent across entire file)
// ---------------------------------------------------------------------------

/// Fake name syllables for generating readable substitutes.
const SYLLABLES: &[&str] = &[
    "ba", "be", "bi", "bo", "bu", "da", "de", "di", "do", "du",
    "fa", "fe", "fi", "fo", "fu", "ga", "ge", "gi", "go", "gu",
    "ka", "ke", "ki", "ko", "ku", "la", "le", "li", "lo", "lu",
    "ma", "me", "mi", "mo", "mu", "na", "ne", "ni", "no", "nu",
    "pa", "pe", "pi", "po", "pu", "ra", "re", "ri", "ro", "ru",
    "sa", "se", "si", "so", "su", "ta", "te", "ti", "to", "tu",
    "va", "ve", "vi", "vo", "vu", "za", "ze", "zi", "zo", "zu",
];

/// Generate a readable fake word from an index.
/// Produces words like "Bako", "Delu", "Fimo", "Gape", etc.
pub fn fake_word_from_index(index: usize) -> String {
    let base = SYLLABLES.len(); // 70 syllables
    let s1 = index % base;
    let s2 = (index / base) % base;

    let mut word = String::new();
    // First syllable capitalized
    let first = SYLLABLES[s1];
    let mut chars = first.chars();
    if let Some(c) = chars.next() {
        word.push(c.to_uppercase().next().unwrap_or(c));
    }
    for c in chars {
        word.push(c);
    }
    // Second syllable
    word.push_str(SYLLABLES[s2]);

    // For very large indices, add a numeric suffix
    let tier = index / (base * base);
    if tier > 0 {
        word.push_str(&tier.to_string());
    }

    word
}

/// Build a global text substitution map from all unique text values.
/// Same text → same fake word everywhere in the file.
/// Returns (original→fake, fake→original) mappings.
pub fn build_global_text_map(
    unique_texts: &[String],
) -> HashMap<String, String> {
    let mut sorted = unique_texts.to_vec();
    sorted.sort();
    sorted.dedup();

    sorted
        .into_iter()
        .enumerate()
        .map(|(i, original)| (original, fake_word_from_index(i)))
        .collect()
}

// ---------------------------------------------------------------------------
// Random parameter generators (matching Python behavior)
// ---------------------------------------------------------------------------

/// Generate a random alphanumeric salt string.
pub fn random_salt(rng: &mut impl Rng, length: usize) -> String {
    const CHARSET: &[u8] = b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    (0..length)
        .map(|_| CHARSET[rng.gen_range(0..CHARSET.len())] as char)
        .collect()
}

/// Return a random scale factor in [0.1, 0.9) ∪ (1.1, 2.0].
pub fn random_factor(rng: &mut impl Rng) -> f64 {
    loop {
        let f: f64 = rng.gen_range(0.1..2.0);
        if !(0.9..=1.1).contains(&f) {
            return (f * 1_000_000.0).round() / 1_000_000.0;
        }
    }
}

/// Return a random numeric offset in [-1000, -100] ∪ [100, 1000].
pub fn random_offset(rng: &mut impl Rng) -> f64 {
    let v: f64 = rng.gen_range(100.0..1000.0);
    let sign = if rng.gen_bool(0.5) { 1.0 } else { -1.0 };
    (v * sign * 10000.0).round() / 10000.0
}

/// Return a random day offset between ±365.
pub fn random_day_offset(rng: &mut impl Rng) -> i32 {
    rng.gen_range(-365..=365)
}

/// Build a deterministic shuffle mapping: sorted unique values → Cat_0, Cat_1, ...
pub fn build_shuffle_mapping(unique_values: &[String]) -> HashMap<String, String> {
    let mut sorted: Vec<String> = unique_values.to_vec();
    sorted.sort();
    sorted.dedup();
    sorted
        .into_iter()
        .enumerate()
        .map(|(i, v)| (v, format!("Cat_{}", i)))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand::rngs::SmallRng;

    #[test]
    fn test_hash() {
        let result = transform_hash("hello", "salt123");
        assert_eq!(result.len(), 12);
        // Same input + salt = same hash
        assert_eq!(result, transform_hash("hello", "salt123"));
        // Different salt = different hash
        assert_ne!(result, transform_hash("hello", "other"));
    }

    #[test]
    fn test_offset_date() {
        let d = NaiveDate::from_ymd_opt(2024, 6, 15).unwrap();
        let shifted = transform_offset_date(d, 10);
        assert_eq!(shifted, NaiveDate::from_ymd_opt(2024, 6, 25).unwrap());
        let restored = reverse_offset_date(shifted, 10);
        assert_eq!(restored, d);
    }

    #[test]
    fn test_offset_number() {
        let v = transform_offset_number(100.0, 50.5);
        assert!((v - 150.5).abs() < 1e-9);
        let r = reverse_offset_number(v, 50.5);
        assert!((r - 100.0).abs() < 1e-9);
    }

    #[test]
    fn test_scale() {
        let v = transform_scale(100.0, 2.5).unwrap();
        assert!((v - 250.0).abs() < 1e-9);
        let r = reverse_scale(v, 2.5).unwrap();
        assert!((r - 100.0).abs() < 1e-9);
        assert!(transform_scale(100.0, 0.0).is_err());
    }

    #[test]
    fn test_shuffle() {
        let mut map = HashMap::new();
        map.insert("USA".to_string(), "Cat_0".to_string());
        map.insert("UK".to_string(), "Cat_1".to_string());
        assert_eq!(transform_shuffle("USA", &map), "Cat_0");
        assert_eq!(transform_shuffle("FR", &map), "FR");
        assert_eq!(reverse_shuffle("Cat_0", &map), "USA");
        assert_eq!(reverse_shuffle("Cat_99", &map), "Cat_99");
    }

    #[test]
    fn test_anonymize() {
        let mut rng = SmallRng::seed_from_u64(42);
        let v = transform_anonymize(&CellValue::Int(100), "int", &mut rng);
        matches!(v, CellValue::Int(_));

        let mut rng = SmallRng::seed_from_u64(42);
        let v = transform_anonymize(&CellValue::String("hello".into()), "string", &mut rng);
        if let CellValue::String(s) = v {
            assert_eq!(s.len(), 5);
            assert!(s.chars().all(|c| c.is_ascii_uppercase()));
        } else {
            panic!("Expected string");
        }
    }

    #[test]
    fn test_build_shuffle_mapping() {
        let vals = vec!["B".into(), "A".into(), "C".into(), "A".into()];
        let map = build_shuffle_mapping(&vals);
        assert_eq!(map.get("A").unwrap(), "Cat_0");
        assert_eq!(map.get("B").unwrap(), "Cat_1");
        assert_eq!(map.get("C").unwrap(), "Cat_2");
    }

    #[test]
    fn test_random_factor_range() {
        let mut rng = SmallRng::seed_from_u64(0);
        for _ in 0..100 {
            let f = random_factor(&mut rng);
            assert!(f >= 0.1 && f <= 2.0);
            assert!(!(0.9..=1.1).contains(&f));
        }
    }
}
